engine/src/test/java/io/zeebe/engine/processor/StreamProcessorTest.java

```diff
@@ -87,6 +87,37 @@ public void onRecovered(final ReadonlyProcessingContext context) {
     inOrder.verifyNoMoreInteractions();
   }
 
+  @Test
+  public void shouldCallStreamProcessorLifecycleOnFail() throws InterruptedException {
+    // given
+    final StreamProcessorLifecycleAware lifecycleAware = mock(StreamProcessorLifecycleAware.class);
+    final CountDownLatch failedLatch = new CountDownLatch(1);
+    streamProcessorRule.startTypedStreamProcessor(
+        (processors, state) ->
+            processors
+                .withListener(lifecycleAware)
+                .withListener(
+                    new StreamProcessorLifecycleAware() {
+                      @Override
+                      public void onRecovered(final ReadonlyProcessingContext context) {
+                        throw new RuntimeException("force fail");
+                      }
+
+                      @Override
+                      public void onFailed() {
+                        failedLatch.countDown();
+                      }
+                    }));
+
+    // when
+    assertThat(failedLatch.await(1000, TimeUnit.MILLISECONDS)).isTrue();
+
+    // then
+    verify(lifecycleAware, times(1)).onRecovered(any());
```

Comment:
do we need the mock since we already have the latch with the other listener ?

Reply:
Found the mock useful to verify that onClose() is never called.

Reply:
Yes I saw that, but actually we should avoid to test something which should not happen, because this is often error prone and hard to test. I mean how long do you want to wait to ensure that it _never_ happens :) It makes also hard to change things later. We should more focus on things we can verify.

Reply:
Make sense. But if we don't test it, what prevents me from accidentally invoking listener:onClose() when the actor failed? Isn't part of the contract that either listener.onFailed or listener.onClose is invoked but not both?

Reply:
Actually nothing. But the question is also should the test prevent you from such thing? When the end result is the same it should be fine right? No matter how many or what kind of methods are called. Is it really part of the contract ? Does it need to be? What values gives us this. As I said we should try to avoid these kind of couplings, since it makes harder to add new features. 

I know I also did this sometimes, since sometimes we have the feeling that we have to make sure such thing doesn't happen. But it makes things often more complicate then it needs to be. We should try to focus on things we can guarantee and verify without using mocks. Like I put `a` into a function and get `b`, but I don't care how it is calculated. On these listeners tests it is a bit harder I know. 

But what we can do is just test things which should happen. Like if something fails it should call `onFailure`. When I close it, it should call `onClose`.

Reply:
ok :+1: